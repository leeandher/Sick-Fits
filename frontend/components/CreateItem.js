import gql from "graphql-tag"
import React, { Component } from "react"
import Form from "./styles/Form"
import { Mutation } from "react-apollo"
import Router from "next/router"

import ErrorMessage from "./ErrorMessage"

import { IMAGE_ENDPOINT } from "../config"

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: "",
    description: "",
    image: "",
    largeImage: "",
    price: 0,
    uploading: false
  }

  handleChange = ({ target: { name, type, value } }) => {
    const stateValue = type === "number" ? parseFloat(value) : value
    this.setState({ [name]: stateValue })
  }

  uploadFile = async e => {
    const [file] = e.target.files
    const data = new FormData()
    data.append("file", file)
    // Select the correct cloudinary upload-preset configuration
    data.append("upload_preset", "sick-fits")
    await this.setState({ uploading: true })
    const res = await fetch(IMAGE_ENDPOINT, {
      method: "POST",
      body: data
    })
    const upload = await res.json()
    if (upload.error) {
      window.alert(upload.error.message)
      return this.setState({ uploading: false })
    }
    console.log(upload)
    this.setState({
      image: upload.secure_url,
      largeImage: upload.eager[0].secure_url,
      uploading: false
    })
  }

  render() {
    const { description, image, price, title, uploading } = this.state
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
            onSubmit={async e => {
              e.preventDefault()
              if (uploading) return false
              // Call the mutation
              const { data } = await createItem()
              // Send them to the single item page
              Router.push({
                pathname: "/item",
                query: { id: data.createItem.id }
              })
            }}
          >
            <ErrorMessage error={error} />
            <fieldset
              disabled={loading || uploading}
              aria-busy={loading || uploading}
            >
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  onChange={this.uploadFile}
                  required
                />
                {image.length > 0 && (
                  <img
                    src={image}
                    alt="Upload Preview"
                    style={{
                      maxWidth: "300px",
                      maxHeight: "200px",
                      padding: "1rem"
                    }}
                  />
                )}
              </label>
              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  value={title}
                  onChange={this.handleChange}
                  required
                />
              </label>
              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  min={0}
                  placeholder="Price"
                  value={price}
                  onChange={this.handleChange}
                  required
                />
              </label>
              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter a Description"
                  value={description}
                  onChange={this.handleChange}
                  style={{ boxShadow: "none" }}
                  required
                />
              </label>
              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
