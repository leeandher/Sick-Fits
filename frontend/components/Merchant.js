import React, { Component } from "react"
import StripeCheckout from "react-stripe-checkout"
import { Mutation } from "react-apollo"
import Router from "next/router"
import gql from "graphql-tag"
import PropTypes from "prop-types"
import NProgress from "nprogress"

import ErrorMessage from "./ErrorMessage"
import User, { CURRENT_USER_QUERY } from "./User"

import calcTotalPrice from "../lib/calcTotalPrice"
import { STRIPE_KEY } from "../config"

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`

class Merchant extends Component {
  totalItems = cart => {
    return cart.reduce((total, cartItem) => total + cartItem.quantity, 0)
  }
  onToken = ({ id }, createOrder) => {
    // Send the ID server-side
    // Manually call the mutation once we have the stripe token
    createOrder({
      variables: { token: id }
    }).catch(err => alert(err.message))
  }
  render() {
    const { children } = this.props
    return (
      <User>
        {({ data: { me } }) => (
          <Mutation
            mutation={CREATE_ORDER_MUTATION}
            refetchQueries={[{ query: CURRENT_USER_QUERY }]}
          >
            {(createOrder, { data }) => (
              <StripeCheckout
                amount={calcTotalPrice(me.cart)}
                name="Sick Fits"
                description={`Order of ${this.totalItems(me.cart)} sick items!`}
                image={me.cart[0].item && me.cart[0].item.image}
                stripeKey={STRIPE_KEY}
                currency="CAD"
                email={me.email}
                token={res => this.onToken(res, createOrder)}
              >
                {children}
              </StripeCheckout>
            )}
          </Mutation>
        )}
      </User>
    )
  }
}

export default Merchant
