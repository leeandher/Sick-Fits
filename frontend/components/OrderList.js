import React, { Component } from "react"
import gql from "graphql-tag"
import styled from "styled-components"
import { Query } from "react-apollo"
import { formatDistance } from "date-fns"
import formatMoney from "../lib/formatMoney"
import Head from "next/head"
import Link from "next/link"
import OrderItemStyles from "./styles/OrderItemStyles"
import ErrorMessage from "./ErrorMessage"

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    orders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`

const ListWrap = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`
const MainWrap = styled.div`
  h2 > span {
    padding: 1rem;
    background: ${({ theme }) => theme.red};
    color: white;
    line-height: 1;
    transform: skew(-6deg);
    display: inline-block;
  }
`

class OrderList extends Component {
  render() {
    return (
      <Query query={USER_ORDERS_QUERY}>
        {({ data: { orders }, error, loading }) => {
          if (loading) return <p>⚡ Loading... ⚡</p>
          if (error) return <ErrorMessage error={error} />
          return (
            <MainWrap>
              <Head>
                <title>Sick Fits - Orders</title>
              </Head>
              <h2>
                You have <span>{orders.length}</span> orders!
              </h2>
              <ListWrap>
                {orders.map(order => (
                  <Link
                    href={{ pathname: "order", query: { id: order.id } }}
                    key={order.id}
                  >
                    <div>
                      <p>
                        <span>Order ID:</span>
                        <span>{order.id}</span>
                      </p>
                      <p>
                        <span>Date:</span>
                        <span>
                          {formatDistance(
                            order.createdAt,
                            "MMMM d, YYY h:mm a"
                          )}
                        </span>
                      </p>
                      <p>
                        <span>Total:</span>
                        <span>{formatMoney(order.total)}</span>
                      </p>
                    </div>
                  </Link>
                ))}
              </ListWrap>
            </MainWrap>
          )
        }}
      </Query>
    )
  }
}

export default OrderList
export { USER_ORDERS_QUERY }
