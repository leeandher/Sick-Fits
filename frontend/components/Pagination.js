import React from 'react'
import PaginationStyles from './styles/PaginationStyles'
import gql from 'graphql-tag'
import { Query } from 'react-apollo'

import { PER_PAGE } from '../config'

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`

const Pagination = ({ page }) => (
  <PaginationStyles>
    <Query query={PAGINATION_QUERY}>
      {({ data, error, loading }) => {
        if (loading) return <p>⚡ Loading... ⚡</p>
        if (error) return <p>❌ Error ❌: {error.message}</p>
        const { count } = data.itemsConnection.aggregate
        const pages = Math.ceil(count / PER_PAGE)
        return (
          <p>
            Page {page} of {pages}
          </p>
        )
      }}
    </Query>
  </PaginationStyles>
)

export default Pagination
