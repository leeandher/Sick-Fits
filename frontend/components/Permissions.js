import { Query } from 'react-apollo'
import gql from 'graphql-tag'

import ErrorMessage from './ErrorMessage'
import Table from './styles/Table'

const ALL_USERS_AND_PERMISSIONS_QUERY = gql`
  query ALL_USERS_AND_PERMISSIONS_QUERY {
    users {
      id
      name
      email
      permissions
    }
    __type(name: "Permission") {
      name
      enumValues {
        name
      }
    }
  }
`

const Permissions = props => (
  <Query query={ALL_USERS_AND_PERMISSIONS_QUERY}>
    {({ data, loading, error }) => {
      const permissionValues = data.__type.enumValues.map(({ name }) => name)
      console.log(permissionValues)
      return (
        <>
          <ErrorMessage error={error} />
          <div>
            <h2>Manage Permissions</h2>
            <Table>
              <thead>
                <tr>
                  <th>Name</th>
                  <th>Email</th>
                </tr>
              </thead>
              <tbody>test</tbody>
            </Table>
          </div>
        </>
      )
    }}
  </Query>
)

export default Permissions
export { ALL_USERS_AND_PERMISSIONS_QUERY }
