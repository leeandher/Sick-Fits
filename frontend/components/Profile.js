import React from 'react'
import styled from 'styled-components'
import { format, formatDistance } from 'date-fns'
import Head from 'next/head'
import formatMoney from '../lib/formatMoney'
import ErrorMessage from './ErrorMessage'
import User from './User'

const ProfileWrapper = styled.div``

const Profile = () => {
  return (
    <User>
      {({ data: { me } }) => (
        <ProfileWrapper>
          <Head>
            <title>Sick Fits | {me.name}</title>
          </Head>
          <p>
            Profile is {formatDistance(Date.now(), Date.parse(me.createdAt))}{' '}
            old
          </p>
          <p>{format(Date.parse(me.createdAt), 'MMMM d, YYY h:mm a')}</p>
          {JSON.stringify(me)}
        </ProfileWrapper>
      )}
    </User>
  )
}

export default Profile
