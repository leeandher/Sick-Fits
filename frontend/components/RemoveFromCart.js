import React, { Component } from "react"
import { Mutation } from "react-apollo"
import gql from "graphql-tag"
import PropTypes from "prop-types"
import styled from "styled-components"

import { CURRENT_USER_QUERY } from "./User"

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${({ theme }) => theme.red};
  }
`

class RemoveFromCart extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired
  }
  render() {
    const { id } = this.props
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id }}
        refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      >
        {(removeFromCart, { loading }) => (
          <BigButton
            title="Delete Item"
            onClick={() => {
              removeFromCart().catch(err => window.alert(err.message))
            }}
            disabled={loading}
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    )
  }
}

export default RemoveFromCart
