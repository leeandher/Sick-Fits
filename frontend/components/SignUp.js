import gql from "graphql-tag"
import React, { Component } from "react"
import { Mutation } from "react-apollo"
import ErrorMessage from "./ErrorMessage"
import Form from "./styles/Form"

const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
  ) {
    signUp(email: $email, name: $name, password: $password) {
      id
      email
      name
    }
  }
`

export default class SignUp extends Component {
  state = {
    email: "",
    name: "",
    password: ""
  }
  saveToState = e => {
    const { value, name } = e.target
    this.setState({ [name]: value })
  }
  render() {
    const { email, name, password } = this.state
    return (
      <Mutation mutation={SIGNUP_MUTATION} variables={this.state}>
        {(signUp, { error, loading }) => {
          return (
            <Form
              method="post"
              onSubmit={async e => {
                e.preventDefault()
                await signUp()
                this.setState({
                  email: "",
                  name: "",
                  password: ""
                })
              }}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Sign Up for an Account</h2>
                <ErrorMessage error={error} />
                <label htmlFor="email">
                  Email
                  <input
                    type="text"
                    name="email"
                    placeholder="e.g. kanye@west.com"
                    value={email}
                    onChange={this.saveToState}
                  />
                </label>
                <label htmlFor="name">
                  Name
                  <input
                    type="text"
                    name="name"
                    placeholder="e.g. Kanye West"
                    value={name}
                    onChange={this.saveToState}
                  />
                </label>
                <label htmlFor="password">
                  Password
                  <input
                    type="password"
                    name="password"
                    placeholder="e.g. y33zy"
                    value={password}
                    onChange={this.saveToState}
                  />
                </label>
                <button type="submit">Sign Up!</button>
              </fieldset>
            </Form>
          )
        }}
      </Mutation>
    )
  }
}
