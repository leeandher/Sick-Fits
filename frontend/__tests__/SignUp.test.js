import { mount } from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import { MockedProvider } from 'react-apollo/test-utils'
import Router from 'next/router'

import SignUp, { SIGN_UP_MUTATION } from '../components/SignUp'
import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser } from '../lib/testUtils'
import { ApolloConsumer } from 'react-apollo'

function type(wrapper, name, value) {
  wrapper.find(`input[name="${name}"]`).simulate('change', {
    target: { name, value },
  })
}

const me = fakeUser()
delete me.createdAt
delete me.updatedAt

const mocks = [
  // Signup mock
  {
    request: {
      query: SIGN_UP_MUTATION,
      variables: {
        name: me.name,
        email: me.email,
        password: 'abc123',
        confirmPassword: 'abc123',
      },
    },
    result: {
      data: {
        signUp: {
          __typename: 'User',
          id: me.id,
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  // Current user mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me } },
  },
]

Router.router = {
  push: jest.fn(),
}

describe('<SignUp />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider>
        <SignUp />
      </MockedProvider>,
    )
    expect(toJSON(wrapper.find('form'))).toMatchSnapshot()
  })
  it('calls the mutation properly', async () => {
    let apolloClient
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client
            return <SignUp />
          }}
        </ApolloConsumer>
      </MockedProvider>,
    )
    await wait()
    wrapper.update()
    // Simulator helper function
    type(wrapper, 'name', me.name)
    type(wrapper, 'email', me.email)
    type(wrapper, 'password', 'abc123')
    type(wrapper, 'confirmPassword', 'abc123')
    wrapper.update()
    wrapper.find('form').simulate('submit')
    await wait()
    // Query the user out of the Apollo Client
    const { data } = await apolloClient.query({ query: CURRENT_USER_QUERY })
    expect(data.me).toMatchObject(me)
  })
})
