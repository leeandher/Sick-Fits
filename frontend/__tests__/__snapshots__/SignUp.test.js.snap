// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp /> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 ZWDQB"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Up for an Account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="e.g. kanye@west.com"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="e.g. Kanye West"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="e.g. y33zy"
        type="password"
        value=""
      />
    </label>
    <label
      htmlFor="confirmPassword"
    >
      Confirm Password
      <input
        name="confirmPassword"
        onChange={[Function]}
        placeholder="e.g. y33zy"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
